Fix widget id stability

AnOctopus opened the pull request:
## Describe your changes
Fix existing widget id instability in slider, document how to safely access session state in widgets, and have `register_widget` check some common fields that shouldn't be used for a widget's identity.

## GitHub Issue Link (if applicable)

## Testing Plan

- Unit Tests (JS and/or Python)
Some additional python unit tests should be added

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

AnOctopus committed:
Add (failing) slider id stability test

AnOctopus committed:
Hacky fix for slider id stability

AnOctopus committed:
Document safe dependency on state in slider

AnOctopus committed:
Crude early detection of common nonidentity fields

AnOctopus committed:
Add new widget id function

AnOctopus committed:
Convert most widgets to new ID generation

AnOctopus committed:
Add docs for requirements for calling widget id gen

AnOctopus committed:
Use current form id as widget id input

AnOctopus committed:
Merge branch 'develop' into fix/widget-id-stability

AnOctopus committed:
Add missing label to time widget id

AnOctopus committed:
Early generate ids for custom components and file uploader

AnOctopus committed:
Generate id for data editor

AnOctopus committed:
Fix and simplify keyed widget id test

AnOctopus committed:
Update other deltagen widget id tests

AnOctopus committed:
Warn about fallback widget id generation

AnOctopus committed:
Concision

AnOctopus committed:
Add widget id kwargs types, start resolving conflicts

AnOctopus committed:
Merge branch 'develop' into fix/widget-id-stability

AnOctopus committed:
Allow common types with safe str encodings

AnOctopus committed:
Finish hacky hashing of existing safe values

AnOctopus committed:
Fix import cycle

AnOctopus committed:
Merge branch 'develop' into fix/widget-id-stability

AnOctopus committed:
Merge branch 'develop' into fix/widget-id-stability

AnOctopus committed:
Clean up union in signature

AnOctopus committed:
Partial processing of time value

AnOctopus committed:
Date parsing

AnOctopus committed:
Hash multiselct index for default

AnOctopus committed:
Add unresolved todo

AnOctopus committed:
Fix dumb errors

vdonato commented on lib/streamlit/runtime/state/common.py:
```diff
@@ -158,6 +175,41 @@ def compute_widget_id(
+    # TODO make sure this is equivalent to the protobuf approach
+    # Values should be of the types that go into protobufs:
+    # float, int, bool, str, bytes, a list of those, or a map of those
+    for k, v in kwargs.items():
```
It'd probably be useful to make a note here that this code relies on Python's dict iteration order being stable, which as far as I understand is designated as something that'll always be true in CPython, but it's not immediately obvious that this is the case.

vdonato commented on lib/streamlit/components/v1/components.py:
```diff
@@ -184,6 +185,26 @@ def marshall_element_args():
 
             if key is None:
                 marshall_element_args()
```
nit: need to update the comment above the definition of `marshall_element_args` to reflect what happens in the new world (it still refers to hashing protobufs)

vdonato commented on lib/streamlit/elements/widgets/data_editor.py:
```diff
@@ -787,7 +791,24 @@ def data_editor(
         # Throws an exception if any of the configured types are incompatible.
         _check_type_compatibilities(data_df, column_config_mapping, dataframe_schema)
 
+        id = new_compute_widget_id(
```
is this something that we want to do closer to the top of the method so that anything that happens in between can't possibly affect the ID? If not, we might want a comment explaining why we defer generating this ID until later

vdonato commented on lib/streamlit/runtime/state/widgets.py:
```diff
@@ -148,12 +151,16 @@ def register_widget(
     """
-    widget_id = compute_widget_id(element_type, element_proto, user_key)
-    element_proto.id = widget_id
+    if not element_proto.id:
+        widget_id = compute_widget_id(element_type, element_proto, user_key)
+        element_proto.id = widget_id
+        LOGGER.warn(
+            f"Registering a widget with no id, falling back to generating from proto. {widget_id=}"
+        )
```
Is this something that we'll want to get rid of before checking this code into `develop`, or are there still instances of us falling back to hashing protos to generate widget IDs? Might want to leave a small comment describing when we still need the fallback if so

AnOctopus replied:
I included it out of caution, in case I had missed something in some edge case, but everything works even replacing it with an assert that id is set, so it shouldn't be needed anymore.

vdonato reviewed (commented):
This LGTM!

I'll take a second pass at the few widget-specific changes in a day or so when I have some more bandwidth, but in terms of the overall approach this is pretty much what I expected things to look like (I'll hit the official +1 after carefully reviewing the more involved widget changes).

Another small thing to note is that there are a few comments spread throughout the codebase that say "This needs to be done after register_widget because we don't want the following proto fields to affect a widget's ID." that are now stale, so we'll want to get rid of all of those.

AnOctopus committed:
Use parse function for min and max

AnOctopus committed:
Rename old widget id computation function

AnOctopus committed:
Rename new widget id function

AnOctopus committed:
Document safety from ensured dict iter order

AnOctopus committed:
Update custom components hashing comments

AnOctopus committed:
Add missing widget id computation do download button

AnOctopus committed:
Remove unneeded fallback id calculation

AnOctopus committed:
Remove old widget id function

AnOctopus committed:
Comment about late id calculation in data_editor

AnOctopus committed:
Update slider session_state use documentation

Most of the caution we needed to instill previously is unnecessary,
because we compute the id before even accessing state.

AnOctopus committed:
Try changing name to unconfuse codeql

AnOctopus committed:
Remove dead imports

AnOctopus committed:
Avoid impossible uninitialized variable

`parsed_value` is always set in the function if `v` has the required
type, but codeql doesn't understand that. Use direct returns so it sees
the impossible case as an implicit None return, which is valid.

kmcgrady commented on lib/streamlit/runtime/state/widgets.py:
```diff
@@ -30,14 +31,15 @@
 if TYPE_CHECKING:
     from streamlit.runtime.scriptrunner import ScriptRunContext
 
+LOGGER = get_logger(__name__)
```
This used somewhere?

AnOctopus committed:
Use `disabled` in data editor id

It has a different meaning than for most widgets.

kmcgrady reviewed (commented):
Do you think it makes sense to write tests for each widget that does the following:

patch `compute_widget_id`

call widget function

assert that compute_widget_id includes all args minus the disallowed for the widget (eg. disabled.

My fear is worrying that a new field is added, but not added to the computer widget id. This test would at least fail here and cause a conversation to be had on whether it should be disallowed or just ensured it's added to the compute_widget_id

AnOctopus committed:
Remove dead code

vdonato commented on lib/streamlit/elements/widgets/button.py:
```diff
@@ -339,16 +340,28 @@ def _download_button(
         check_session_state_rules(default_value=None, key=key, writes_allowed=False)
+
+        id = compute_widget_id(
+            "button",
```
nit: should this type be `download_button`?

Also heads up that this change may cause some weird behavioral issues with #7056 that may not show up as merge conflicts.

vdonato reviewed (approved):
LGTM

AnOctopus commented:
> Do you think it makes sense to write tests for each widget that does the following:
> 
> patch `compute_widget_id`
> 
> call widget function
> 
> assert that compute_widget_id includes all args minus the disallowed for the widget (eg. disabled.
> 
> My fear is worrying that a new field is added, but not added to the computer widget id. This test would at least fail here and cause a conversation to be had on whether it should be disallowed or just ensured it's added to the compute_widget_id

We could write those tests, but it seems annoying, and I'm not sure it adds that much value. I expect it to be easy to remember to update the call, and any e2e test that has multiple copies of a widget with just that argument differing will get a duplicate id error (though maybe that won't happen because they'll have different labels? unsure).

Even if we decide it would be worth having those tests, I don't want the PR to be blocked on them. I'd rather they be a quick followup.

AnOctopus committed:
Distinguish download button id

AnOctopus commented:
> LGTM

Cool. I'll wait to merge it until tomorrow, to give Lukas a chance to do a closer check of the data editor id calculation, since I want his verification that we're using the arrow table.

vdonato commented:
> We could write those tests, but it seems annoying, and I'm not sure it adds that much value. I expect it to be easy to remember to update the call, and any e2e test that has multiple copies of a widget with just that argument differing will get a duplicate id error (though maybe that won't happen because they'll have different labels? unsure).

I think we won't get any e2e test coverage for this since most of our tests have widgets with different labels.

There's probably a clever way to write these tests without ridiculous amounts of code duplication using `inspect.signature` on the widget functions and asserting that the kwargs that `compute_widget_id` is called with match (with the exception of the few parameters that we don't expect to be included in the widget ID calc). I'd probably be okay with that being a quick followup PR after this one is checked in, though, unless the `inspect.signature` trick makes writing these tests not too difficult.

lukasmasuch commented on lib/streamlit/elements/widgets/data_editor.py:
```diff
@@ -787,7 +791,29 @@ def data_editor(
+            "data_editor",
+            user_key=key,
+            # TODO is this in a format where the string bytes accurately represent it?
+            data=arrow_table,
```
To be safe, I think we can move the `pyarrow_table_to_bytes` call above the id generation and directly use the bytes here:

```suggestion
        arrow_bytes = type_util.pyarrow_table_to_bytes(arrow_table)
        
        # We want to do this as early as possible to avoid introducing nondeterminism,
        # but it isn't clear how much processing is needed to have the data in a
        # format that will hash consistently, so we do it late here to have it
        # as close as possible to how it used to be.
        id = compute_widget_id(
            "data_editor",
            user_key=key,
            data=arrow_bytes,
```

lukasmasuch commented on lib/streamlit/elements/widgets/data_editor.py:
```diff
@@ -787,7 +791,29 @@ def data_editor(
+            column_order=column_order,
+            column_config=str(column_config),
+            num_rows=num_rows,
+            disabled=(disabled if isinstance(disabled, bool) else list(disabled)),
```
Maybe using the `column_config_mapping` is a bit better than the raw input parameter since it is closer to what is stored in the proto. This also allows leaving out the `hide_index` and `disabled` parameters since these ar already baked into the` column_config_mapping`. Also, if `disabled` is a boolean we probably don't want to have it influence the id. It's not how the current impl. does it (I probably didn't think about that), but I think it would be closer to how other widgets handle `disabled`. 

```suggestion
            column_order=column_order,
            column_config_mapping=str(column_config_mapping),
            num_rows=num_rows,
```

lukasmasuch commented on lib/streamlit/elements/widgets/chat.py:
```diff
@@ -277,6 +278,13 @@ def chat_input(
         check_callback_rules(self.dg, on_submit)
         check_session_state_rules(default_value=default, key=key, writes_allowed=False)
 
+        id = compute_widget_id(
```
A related question here: we have [a request](https://github.com/streamlit/streamlit/issues/7054) to allow changing the `placeholder` without resetting the widget itself. I thought that we might be able to allow this use-case in case the `key` is provided by the user, e.g.: 

```python
if key:
    id = compute_widget_id(
        "chat_input",
        user_key=key,
        max_chars=max_chars,
    )
else:
    id = compute_widget_id(
        "chat_input",
        user_key=None,
        placeholder=placeholder,
        max_chars=max_chars,
    )
```

Do you think this work, or are there any negative side-effects of doing this?

AnOctopus replied:
Let's keep that to a separate PR. Doing it like that is fine on the technical side (custom components now does it this way), the question is if we should do it, which is a UX/product design matter.

lukasmasuch reviewed (approved):
Overall LGTM 👍  I added some suggestions for the data editor implementation to make it a bit more similar to the data used for ID generation in the current version.

AnOctopus committed:
Use column_config_mapping

AnOctopus committed:
Use arrow_bytes
